### ISSUE TITLE
Add more examples to documentation (concrete wish: TokenFilter-Example)

### ISSUE THREAD

**Issue by klartext** [pikepdf/pikepdf#62]
I would like to remove objects from a pdf with certain properties (to be precise: remove curve-objects of a certain size from the pdf and save the reduced pdf in a seperate file).
I found no example for theTokenFilter in the documentation.
So I ask for such examples.
Of course other sections might also benefit from more examples.

Links to projects that use pikepdf would also make sense as examples and it would make sense to add such links to the documentation.

**Comment by jbarlow83** [comment]
I agree it's not adequate. Issue is just the time to write documentation.

I'd use a content stream parser for suppressing commands. The token filter is very low level and not aware of the semantics of tokens. The parser will tell you if a given command is an operator.

It would be to remove curves because you need to work out the full graphics state at the end of the operation. If the curve is used for something fancy like a clipping region then removing it may have odd side effects. (Unless the curve is neatly packaged inside a `q`/`Q` operator pair.)

**Comment by klartext** [comment]
From the documentation:

"The parser is useful when one wants to
retrieve information from a content stream, such as determine the position of an element. The parser should not be used to edit or reconstruct the content stream because some subtleties are lost in parsing."
and
"The token ﬁlter works at a lower level, considering each token including comments, and distinguishing different types of spaces. This allows modifying content streams."

I read that as: if I want to drop contents (modifiyng streams) I have to use the TokenFilter.

**Comment by klartext** [comment]
Possibly I misunderstood that part of the doecumentation? So, would changing the content stream mean, change the definition of a curve itself, instead of just deleting it as a whole? I try with content stream parser. Is the code you used to suppress commands available somewhere? (Possibly shifting at least a part of it it into the documentation)?

**Comment by jbarlow83** [comment]
You could have a look at the qpdf documentation since pikepdf is providing a fairly thin wrapper for qpdf's features in this area.

You can use a content stream parser to delete a curve. There's no command to suppress. It works like this (from ocrmypdf): we suppress text that is drawn while the render mode is set to `3 Tr` (invisible text) by not appending it to the list of objects, then rebuild the content stream from the list of objects that will be passed through.

```python

def strip_invisible_text(pdf, page):
    stream = []
    in_text_obj = False
    render_mode = 0
    text_objects = []

    page.page_contents_coalesce()
    for operands, operator in pikepdf.parse_content_stream(page, ''):
        if not in_text_obj:
            if operator == pikepdf.Operator('BT'):
                in_text_obj = True
                render_mode = 0
                text_objects.append((operands, operator))
            else:
                stream.append((operands, operator))
        else:
            if operator == pikepdf.Operator('Tr'):
                render_mode = operands[0]
            text_objects.append((operands, operator))
            if operator == pikepdf.Operator('ET'):
                in_text_obj = False
                if render_mode != 3:
                    stream.extend(text_objects)
                text_objects.clear()


    def convert(op):
        try:
            return op.unparse()
        except AttributeError:
            return str(op).encode('ascii')

    lines = []

    for operands, operator in stream:
        if operator == pikepdf.Operator('INLINE IMAGE'):
            iim = operands[0]
            line = iim.unparse()
        else:
            line = b' '.join(convert(op) for op in operands) + b' ' + operator.unparse()
        lines.append(line)

    content_stream = b'\n'.join(lines)
    page.Contents = pikepdf.Stream(pdf, content_stream)

```

**Comment by jbarlow83** [comment]
I added some improvements to read content stream parsing which should be more helpful to most users than TokenFilters. v1.9.0

**Comment by tomifarago** [comment]
Hey!
Can you help me edit the content stream so that I can replace an image with text?
Thanks

**Comment by jbarlow83** [comment]
I recommend using a PDF internals viewer like Acrobat's Preflight feature to compare how text and images are rendered - that will show you could make the changes. 

This is like changing text in HTML  to an image. If the HTML is complex you may need to do fairly complex parsing (up to full layout rendering and JavaScript processing) to find where the text is and where to insert the image.
